### PR TITLE
Use NAPI, Add TS Types, Use Yarn

### PR DESCRIPTION
**Pull request by Robinlemon**
- Uses git submodules instead of a static build
- Adds typescript types
- Uses yarn instead of npm for quicker installs and module cleanup
- Adds docker for clean build testing
- Adds node-gyp config for building the native addon
- Replicates tests in ConfusableMatcher

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# TypeScript NAPI Interop for Confusable Matcher
+
+This library exports a wrapper class bundled with TypeScript declarations that allows you to use the C++ ConfusableMatcher inside your JavaScript (NodeJS/CommonJS) applications.
+
+## Benchmark
+
+```md
+yarn benchmark
+
+ConfusableMatcher#indexOf x 137,296 ops/sec ±0.42% (90 runs sampled)
```
Comment (TETYYS) [inline README.md:10]: missing hardware that this was run at

File: docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+version: '3.7'
+
+services:
+    confusablematcher-js-interop-build-clean:
+        image: moderatex/confusablematcher-js-interop-build-runner:latest
```
Comment (TETYYS) [inline docker-compose.yaml:5]: `moderatex` - unrelated references to consumer product

File: src/interop.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    delete this->_instance;
+}
+
+Napi::Value ConfusableMatcherNapiInterop::AddMapping(const Napi::CallbackInfo &info)
+{
+    Napi::Env env = info.Env();
+    Napi::String Key = info[0].As<Napi::String>();
+    Napi::String Value = info[1].As<Napi::String>();
+    Napi::Boolean CheckValueDuplicate = info[2].As<Napi::Boolean>();
+    bool result = this->_instance->AddMapping(Key.Utf8Value(), Value.Utf8Value(), CheckValueDuplicate.ToBoolean());
```
Comment (TETYYS) [inline src/interop.cpp:43]: This is not thread safe if called during `IndexOf` call

File: src/interop.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    return Napi::Boolean::New(env, result);
+}
+
+Napi::Value ConfusableMatcherNapiInterop::AddSkip(const Napi::CallbackInfo &info)
+{
+    Napi::Env env = info.Env();
+    Napi::String In = info[0].As<Napi::String>();
+    bool result = this->_instance->AddSkip(In.Utf8Value());
```
Comment (TETYYS) [inline src/interop.cpp:51]: This is not thread safe if called during IndexOf call

File: src/interop.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    Napi::Boolean MatchOnWordBoundary = Options.Get("matchOnWordBoundary").As<Napi::Boolean>();
+
+    CMOptions opts = {};
+    opts.MatchRepeating = MatchRepeating.ToBoolean();
+    opts.StartIndex = StartIndex.ToNumber().Int32Value();
+    opts.StartFromEnd = StartFromEnd.ToBoolean();
+    opts.StatePushLimit = StatePushLimit.ToNumber().Int32Value();
+    opts.MatchOnWordBoundary = MatchOnWordBoundary.ToBoolean();
+
+    std::pair<int, int> result = this->_instance->IndexOf(
```
Comment (TETYYS) [inline src/interop.cpp:94]: Fine on current submodule hash, but probably should update

File: src/wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+            throw new Error('Key cannot begin with \\x00 or \\x01');
+        }
+        if (value.length === 0) {
+            throw new Error('Value provided is empty.');
+        }
+        if (value[0] == '\x00' || value[0] == '\x01') {
+            throw new Error('Value cannot begin with \\x00 or \\x01');
+        }
+
+        key = key.toLocaleUpperCase();
```
Comment (TETYYS) [inline src/wrapper.ts:88]: Not required as a part of the lib, this should be handled by consumer

File: src/interop.cpp (hunk 3 above)
Comment (Robinlemon) [reply src/interop.cpp]: Removed ref to internal method

File: src/interop.cpp (hunk 4 above)
Comment (Robinlemon) [reply src/interop.cpp]: Removed ref to internal method

**Pull request merged**